spin/cmds/meson.py

```diff
@@ -319,12 +319,23 @@ def build(
             )
         return
 
-    meson_args = list(meson_args)
+    if isinstance(meson_args, tuple):
+        # for backwards compatibility because
+        # previously meson_args was a tuple.
+        # all the members of the meson_args tuple
+        # were passed to meson setup subcommand.
+        meson_args_ = {}
+        meson_args_["setup"] = meson_args
+        meson_args_["compile"] = tuple()
+        meson_args_["install"] = tuple()
```

Comment:
I don't see these being used below, so do we need the complication of splitting meson_args into three parts?

Reply:
OK, thanks for clarifying. I expected to see some changes to the click parameters to allow specifying such separate sets of settings, but I don't?

Reply:
Yes. It wasn’t needed.

Reply:
So what cli usage will correctly populate this parameter?

Reply:
In SciPy, the code in `spin/.cmds.py`, we add code to figure out which CLI argument should go into which subcommand. If you don't do it, then the code above will be activated and everything will be passed to `meson setup` subcommand.

Reply:
Let me explain more. So in SciPy, we have some SciPy specific build options which are passed via CLI. Now these options are mapped to `meson` arguments. `meson` is called here in `spin`. Some SciPy build options are forwarded to `meson setup` subcommand, some to `meson compile` and some to `meson install`. That's why to need to figure out which SciPy build option will go into `setup`, which ones will go into `compile` and which ones will go into `install`. Therefore this segregation can only be done in SciPy. Same can be happen with any other project, they can define their own build options which will map to some `meson` arguments. They will segregate those options into `setup`, `compile` and `install`. The `spin` code here in this PR will generate `meson` sub-commands and attach those forwarded options to those subcommands.

Let me know if I misinterpreted anything.

Reply:
See https://github.com/scipy/scipy/blob/20f8f5eca1f90c49ffe1133c5af3b22cba31fe65/.spin/cmds.py#L345-L406 to see how the segregation of SciPy build options into meson subcommads is happening.

Reply:
Basically it's about how `meson` is being called in `spin`. If I want to pass a meson argument to `meson install`, then `spin` won't do it. All the `meson_args` are attached to `meson setup` subcommand.

Let me know if I am wrong here.

Reply:
> Then after updating scipy's spin config to use the new extension mechanism added in #248 all projects using spin and meson will get the feature "for free".

I am not sure if the issue I am solving here is linked to #248. I will go through it and see if it can be used here as well.

Reply:
Is there a reason other projects would not benefit from being able to specify those three settings separately? If they would, we could add flags to allow their specification?

If not, I think best to split this into three keyword arguments. That would still allow SciPy to do what it needs, but it's less obscure from the spin perspective (the implicit dictionary structure being used now).

Reply:
> Is there a reason other projects would not benefit from being able to specify those three settings separately? If they would, we could add flags to allow their specification?

Yes, being able to pass arguments to `meson compile` and `meson install` is potentially useful to any project that uses Meson.

This PR in its current form does allow that. If it's easier/nicer to add 3 new keywords rather than mash it into `meson_args`, then sure let's do that.

Reply:
Made the changes as discussed here.